shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,159 @@
+import 'dart:async';
+import 'dart:ffi';
+// import 'dart:ffi';
+import 'dart:isolate';
+
+// import 'package:ffi/ffi.dart';
+import 'package:ffi/ffi.dart';
+import 'package:meta/meta.dart';
+import 'package:shorebird_code_push/src/generated/updater_bindings.g.dart';
+import 'package:shorebird_code_push/src/shorebird_updater.dart';
+import 'package:shorebird_code_push/src/updater.dart';
+
+@visibleForTesting
+
+/// Type definition for [Isolate.run].
+typedef IsolateRun = Future<R> Function<R>(
+  FutureOr<R> Function(), {
+  String? debugName,
+});
+
+/// {@template shorebird_updater_io}
+/// The Shorebird web updater.
+/// {@endtemplate}
+class ShorebirdUpdaterImpl implements ShorebirdUpdater {
+  /// {@macro shorebird_updater_io}
+  ShorebirdUpdaterImpl(this._updater, {IsolateRun? run})
+      : _run = run ?? Isolate.run {
+    try {
+      // If the Shorebird Engine is not available, this will throw an exception.
+      _updater.currentPatchNumber();
+      _isAvailable = true;
+    } catch (_) {
+      logShorebirdEngineUnavailableMessage();
+      _isAvailable = false;
+    }
+  }
+
+  late final bool _isAvailable;
+
+  final Updater _updater;
+
+  final IsolateRun _run;
+
+  @override
+  bool get isAvailable => _isAvailable;
+
+  @override
+  Future<Patch?> readCurrentPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.currentPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<Patch?> readNextPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.nextPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<UpdateStatus> checkForUpdate() async {
+    if (!_isAvailable) return UpdateStatus.unavailable;
+
+    final isUpdateAvailable = await _run(_updater.checkForUpdate);
+    if (isUpdateAvailable) return UpdateStatus.outdated;
+
+    final (current, next) = await (readCurrentPatch(), readNextPatch()).wait;
+    return next != null && current?.number != next.number
+        ? UpdateStatus.restartRequired
+        : UpdateStatus.upToDate;
```

Comment:
Why is this needed?  Should _updater.checkForUpdate return more of what you need?  (I guess that woudl be another breaking chagne).

Reply:
Yeah it's just more breaking rust changes 🤷

Reply:
Since neither of those hit the network I think it's fine.